modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1245 @@
+        nSegmentElems[locIdx] = nElems;
+        locIdx += 1;
+      }
+
+      // Find the average and the excess. The excess is calculated as the amount
+      // of elements a segment has over the average, which is used to calculate
+      // the buffer size for each segment.
+      var total = (+ reduce nSegmentElems);
+      var avg = total / locIdx;
+      var excess : int;
+      for nElems in nSegmentElems {
+        if nElems > avg {
+          excess += nElems - avg;
+        }
+      }
+
+      // Allocate buffer, which holds the 'excess' elements for redistribution.
+      // Then fill it.
+      var buffer = c_malloc(eltType, excess);
+      var bufferOffset = 0;
+      for loc in localThis.targetLocales do on loc {
+        var average = avg;
+        ref segment = getPrivatizedThis.bag.segments[segmentIdx];
+        var nElems = segment.nElems.read() : int;
+        if nElems > average {
+          var take = nElems - average;
+          var tmpBuffer = __primitive("+", buffer, bufferOffset);
+          segment.transferElements(tmpBuffer, take, buffer.locale.id);
+          bufferOffset += take;
+        }
+      }
+
+      // With the excess elements, redistribute it...
+      // BUG: Cannot seem to actually move data this way...
+      bufferOffset = 0;
+      for loc in localThis.targetLocales do on loc {
+        var average = avg;
+        ref segment = getPrivatizedThis.bag.segments[segmentIdx];
+        var nElems = segment.nElems.read() : int;
+        if average > nElems {
+          var give = average - nElems;
+          var arr : [1..give] eltType;
+          on bufferOffset {
+            var tmpBuffer = buffer;
+            for i in 1 .. give {
+              arr[i] = tmpBuffer[bufferOffset];
+              bufferOffset += 1;
+            }
+          }
+          for i in 1 .. give {
+            segment.addElements(arr[i]);
+          }
+        }
+      }
+
+      // Lastly, if there are items left over, just add them to our locale's segment.
+      if excess > bufferOffset {
+        ref segment = localThis.bag.segments[segmentIdx];
+        var give = excess - bufferOffset;
+        var tmpBuffer = __primitive("+", buffer, bufferOffset);
+        segment.addElementsPtr(tmpBuffer, give, buffer.locale.id);
+      }
+
+      c_free(buffer);
+    }
+
+    // Phase 3: Release all locks from first node and segment to last node and segment.
+    for loc in localThis.targetLocales do on loc {
+      var instance = getPrivatizedThis;
+      for segmentIdx in 0 .. #here.maxTaskPar {
+        ref segment = instance.bag.segments[segmentIdx];
+        segment.releaseStatus();
+      }
+    }
+  }
+
+  /*
+    Iterate over each bag in each node. If the bag is frozen, we elide the need
+    to acquire any locks and iterate directly over the bag. If the bag is not
+    frozen, we must acquire not only the needed locks, but also take a snapshot
+    approach, increasing memory consumption.
+  */
+  iter these() : eltType {

```

Comment:
Technically, a lot of things that use remote execution, such as remote atomic operations without network atomics would violate an agreement of 'not using `on` statements'. I'm all for it, especially because you cannot remotely use a C_Ptr and must an `on` statement to make use of it.